File: LSpredictor.m
function [result] = LSpredictor(x, y, ref_value)
    % xy=x-y;
    % d=abs(xy(1)-xy(2))+abs(xy(1)-xy(3))+abs(xy(2)-xy(3));
    % d=x-y;
    %     if ( abs(d(1)-d(2))+abs(d(1)-d(3))+abs(d(2)-d(3)) )>( abs(y(1)-y(2))+abs(y(1)-y(3))+abs(y(2)-y(3)) )
    %         pred_value=MED(y);
    %     else
    %     d=abs(ref_value-MED(x));
    n = length(x);
    nn = min(length(unique(x)), length(unique(y)));
    k = max(nn - 1, 0);
    X = (zeros(n, k + 1));
    X0 = (zeros(1, k + 1));

    for i = 1:n

        for j = 1:k + 1
            X(i, j) = x(i)^(j - 1);

            if i == 1
                X0(1, j) = ref_value^(j - 1);
            end

        end

    end

    A = (X' * X)^-1 * X' * y';
    %         d=round(10*sum(abs(X*A-y')));
    pred_value = round(X0 * A);
    result = pred_value;

    if pred_value > max([y(1) y(2)])
        result = max([y(1) y(2)]);
    end

    if pred_value < min([y(1) y(2)])
        result = min([y(1) y(2)]);
    end

    %      end

end

File: LSpredictor.py
import numpy as np


'''
代码涉及部分
Ⅲ. PROPOSED SCHEME
B. Polynomial Predictor Based on Gray Version(基于灰色版本的多项式预测器)

生成一个具有清晰直方图的序列作为RDH的宿主序列
'''

def LSpredictor(x, y, ref_value):
    n = len(x)
    nn = min(len(np.unique(x)), len(np.unique(y)))
    k = max(nn - 1, 0)
    X = np.zeros([n, k + 1])
    X0 = np.zeros(k + 1)
    for i in range(n):
        for j in range(k + 1):
            X[i][j] = x[i] ** j
            if i == 0:
                X0[j] = ref_value ** j
    
    # 找到最佳多项式系数
    A = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(np.array([y]).T)
    # 将预测结果限制再r(i+1,j)和r(i,j+1)之间
    pred_value = np.round(X0.dot(A))
    result = pred_value[0]
    if pred_value > max(y[0], y[1]):
        result = max(y[0], y[1])
    if pred_value < min(y[0], y[1]):
        result = min(y[0], y[1])
    return result
